contracts/v3-migration/OriginationControllerMigrate.sol

```diff
@@ -0,0 +1,472 @@
+        }
+
+        // collateral address and id must be the same
+        if (sourceLoanTerms.collateralAddress != newLoanTerms.collateralAddress || sourceLoanTerms.collateralId != newLoanTerms.collateralId) {
+            revert OC_RolloverCollateralMismatch(
+                sourceLoanTerms.collateralAddress,
+                sourceLoanTerms.collateralId,
+                newLoanTerms.collateralAddress,
+                newLoanTerms.collateralId
+            );
+        }
+
+        // ------------- New LoanTerms Validation -------------
+        // principal must be greater than or equal to the configured minimum
+        if (newLoanTerms.principal < allowedCurrencies[newLoanTerms.payableCurrency].minPrincipal) revert OC_PrincipalTooLow(newLoanTerms.principal);
+
+        // loan duration must be greater or equal to 1 hr and less or equal to 3 years
+        if (newLoanTerms.durationSecs < 3600 || newLoanTerms.durationSecs > 94_608_000) revert OC_LoanDuration(newLoanTerms.durationSecs);
+
+        // interest rate must be greater than or equal to 0.01% and less or equal to 1,000,000%
+        if (newLoanTerms.interestRate < 1 || newLoanTerms.interestRate > 1e8) revert OC_InterestRate(newLoanTerms.interestRate);
+
+        // signature must not have already expired
+        if (newLoanTerms.deadline < block.timestamp) revert OC_SignatureIsExpired(newLoanTerms.deadline);
+    }
+
+    // ========================================= HELPERS ================================================
+
+    /**
+     * @notice Helper function to distribute funds based on the migration amounts. If the lender is the
+     *         same as the old lender, the flash loan trigger is set to true. This informs the calling
+     *         function that a flash loan must be initiated to repay the old loan.
+     *
+     * @param oldLoanId                 The ID of the v3 loan to be migrated.
+     * @param oldLoanData               The loan data of the v3 loan.
+     * @param newPrincipalAmount        The principal amount of the new loan.
+     * @param borrower_                 The address of the borrower.
+     * @param lender                    The address of the new lender.
+     *
+     * @return amounts                  The migration amounts.
+     * @return flashLoanTrigger         boolean indicating if a flash loan must be initiated.
+     */
+    function _migrate(
+        uint256 oldLoanId,
+        LoanLibraryV3.LoanData memory oldLoanData,
+        uint256 newPrincipalAmount,
+        address borrower_,
+        address lender
+    ) internal nonReentrant returns (OriginationLibrary.RolloverAmounts memory amounts, bool flashLoanTrigger) {
+        address oldLender = ILoanCoreV3(loanCoreV3).lenderNote().ownerOf(oldLoanId);
+        IERC20 payableCurrency = IERC20(oldLoanData.terms.payableCurrency);
+
+        // Calculate settle amounts
+        (amounts) = _calculateV3MigrationAmounts(
```

Comment:
If we use a flash loan, do we need to also add the flash loan fee to `amounts` here?

Reply:
Yes, on second thought I should add this just in case the flash loan provider turns fees on. I added the logic to the flash loan callback. The amount will just have to be approved up front if there are flash loan fees present.